add support for do concurrent - 2nd try (closes #403)

rupertford opened the pull request:
The f2008 `do concurrent` functionality has already been added but does not get called as the original f2003 classes get called directly from other f2003 classes. In this PR, 2008 versions of these other f2003 classes are added so all should be well.

rupertford commented:
Ready for first review from @arporter or @sergisiso

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -7787,23 +7787,35 @@ class Block_Label_Do_Construct(BlockBase):  # pylint: disable=invalid-name
+        :rtype: :py:class:`fparser.two.Fortran2003.Label_Do_Stmt`
+
+        """
+        # Import here to avoid circular imports
```
Please disable the pylint warning.

rupertford replied:
I'm not sure why the import was added as the class is in the same file. Removed.

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -7837,23 +7849,35 @@ class Block_Nonlabel_Do_Construct(BlockBase):  # pylint: disable=invalid-name
+        :rtype: :py:class:`fparser.two.Fortran2003.Nonlabel_Do_Stmt`
+
+        """
+        # Import here to avoid circular imports
```
pylint disable please.

rupertford replied:
As above, not sure why this was added. Removed.

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -7898,9 +7922,21 @@ def match(string):
+        :rtype: :py:class:`fparser.two.Fortran2003.Loop_Control`
+
+        """
+        # Import here to avoid circular imports
```
pylint disable please.

rupertford replied:
Removed import.

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -7947,14 +7983,26 @@ class Nonlabel_Do_Stmt(StmtBase, WORDClsBase):  # pylint: disable=invalid-name
         """
         :param str string: Fortran code to check for a match
         :return: code line matching the nonlabeled "DO" statement
         :rtype: string
```
Not yours but s/string/str/ please.

rupertford replied:
Hopefully you meant for the return type as that is what I've changed.

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -7947,14 +7983,26 @@ class Nonlabel_Do_Stmt(StmtBase, WORDClsBase):  # pylint: disable=invalid-name
+
+        """
+        # Import here to avoid circular imports
+        from fparser.two.Fortran2003 import Loop_Control
```
pylint...

rupertford replied:
Removed as import is not needed.

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -8158,17 +8206,29 @@ class Action_Term_Do_Construct(BlockBase):  # R836
+
+        """
+        # Import here to avoid circular imports
+        from fparser.two.Fortran2003 import Label_Do_Stmt
```
pylint..

rupertford replied:
Removed as not required.

arporter commented on src/fparser/two/Fortran2008/action_term_do_construct_r824.py:
```diff
@@ -0,0 +1,65 @@
+    def label_do_stmt_cls():
+        """
+        :returns: Fortran2008 Label_Do_Stmt class.
+        :rtype: :py:class:`fparser.two.Fortran2003.Label_Do_Stmt`
```
2008

rupertford replied:
Oops.

arporter commented on src/fparser/two/Fortran2008/nonlabel_do_stmt_r817.py:
```diff
@@ -0,0 +1,60 @@
+    def loop_control_cls():
+        """
+        :returns: Fortran2003 Loop_Control class.
+        :rtype: :py:class:`fparser.two.Fortran2003.Loop_Control`
```
2008 here and on line above.

arporter reviewed (changes requested):
Nicely done.
There are a few pylint things to clean-up and I think we could do with one or more slightly more 'functional' tests where we check that 'do concurrent' is parsed correctly in the context of a whole routine - i.e. as a child of a routine, and if-block, another loop...

rupertford commented on src/fparser/two/tests/fortran2008/test_action_term_do_construct_r824.py:
```diff
@@ -0,0 +1,58 @@
+@pytest.mark.usefixtures("f2008_create")
+def test_concurrent():
+    """Test that the Fortran2008 version supports do concurrent."""
+    code = "DO 10 CONCURRENT (i = 1 : 20)\n" "  a(i) = 0.0\n" "10 b(i) = 1.0"
```
Nor do I. Not sure what happend there.

arporter replied:
It's still there...?

rupertford replied:
It was black reformatting the code that way. I've found a solution by making it a single line string so both black and pylint are happy.

rupertford commented:
Ready for next review from @arporter

rupertford commented:
Perhaps the rule should be that any classes that simply subclass another class in a different API so that subsequent classes use the the new API, should have functional tests? I think that would then cover all cases and more. I've not done this in this PR but would be happy to do so if the reviewer thinks this is a better solution than the single functional test I've added for one case when using concurrent.

arporter commented:
> Perhaps the rule should be that any classes that simply subclass another class in a different API so that subsequent classes use the the new API, should have functional tests? I think that would then cover all cases and more. I've not done this in this PR but would be happy to do so if the reviewer thinks this is a better solution than the single functional test I've added for one case when using concurrent.

Something like that, yes. (I don't quite follow what you mean.) In this case I'd be happier if we at least check for the case where the DO CONCURRENT is a child of:
 * a normal DO loop
 * an IF block
 * a SELECT CASE

rupertford commented:
I've added functional tests where the modified classes are purely there to connect the f2003 to the f2008 concurrent classes. I've just added basic functional tests as everything else will work fine, the important thing is that the different flavours of f2008 concurrent can be parsed from f2003 classes (I've just used program).

Ready for next review from @arporter

arporter reviewed (approved):
Thanks for making those changes.
All good now.
Will proceed to merge.
